FTN/ring_buffer.c

```diff
@@ -0,0 +1,167 @@
+#include "./ring_buffer.h"
+
+bool init_ring_buffer(ring_buffer *ring_buff)
+{
+    ring_buff->head = 0;
+    ring_buff->tail = 0;
+    ring_buff->count = 0;
+    ring_buff->head_in_prograss = 0;
+    ring_buff->tail_in_prograss = 0;
+    ring_buff->len_in_prograss = 0;
+    return true;
+}
+
+bool init_ringbuffer_arr(ring_buffer *ring_buff_arr, uint64_t len)
+{
+    uint64_t x = 0;
+    for (x = SERVER_ADDRESS_ID; x < len; x++)
+    {
+        if (!init_ring_buffer(&ring_buff_arr[x]))
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool RB_insert(ring_buffer *ring_buff, msg *new_msg)
+{
+    int64_t index_to_write = 0;
+    uint64_t actual_count = __atomic_add_fetch(&ring_buff->len_in_prograss, 1, __ATOMIC_SEQ_CST);
+    if (actual_count > RING_BUFFER_SIZE)
+    {
+        actual_count = __atomic_fetch_sub(&ring_buff->len_in_prograss, 1, __ATOMIC_SEQ_CST);
+        return 0;
+    }
+    index_to_write = __atomic_fetch_add(&(ring_buff->tail_in_prograss), 1, __ATOMIC_SEQ_CST);
+    memcpy(&(ring_buff->msgs[(index_to_write % RING_BUFFER_SIZE)]), new_msg, sizeof(msg));
+    while ((ring_buff->tail) % RING_BUFFER_SIZE != (index_to_write % RING_BUFFER_SIZE))
```

Comment:
On this code, each writer should wait for all other writers...
Can you suggest a more elegant way that eliminates this issue? (Or at least reduce the number of writers waiting?)